From c28a677a17eeb96042bc126ec5c73bf8eb291fc9 Mon Sep 17 00:00:00 2001
From: Haihua Hu <b55597@freescale.com>
Date: Thu, 25 Feb 2016 13:53:20 +0800
Subject: [PATCH 1/2] [glplugin] glcolorconvert: convert YUV to RGB use
 directviv

1.Modify glcolorconvert to let src pad use proposed buffer pool
2.Add a property "disable_passthrough" in glcolorconvert for enable/disable passthrough.
  When need convert YUV to RGB with directviv, set it to be TRUE.

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Haihua Hu <b55597@freescale.com>
---
 ext/gl/gstglcolorconvertelement.c   |   83 ++++++++++++++++++++++++++++++++++-
 ext/gl/gstglcolorconvertelement.h   |    1 +
 gst-libs/gst/gl/gstglcolorconvert.c |   23 ++++++----
 3 files changed, 97 insertions(+), 10 deletions(-)

diff --git a/ext/gl/gstglcolorconvertelement.c b/ext/gl/gstglcolorconvertelement.c
index e4ff6ca..854edc2 100644
--- a/ext/gl/gstglcolorconvertelement.c
+++ b/ext/gl/gstglcolorconvertelement.c
@@ -35,6 +35,14 @@ G_DEFINE_TYPE_WITH_CODE (GstGLColorConvertElement, gst_gl_color_convert_element,
         "glconvertelement", 0, "convert");
     );
 
+enum
+{
+  GL_COLOR_CONVERT_PROP_0,
+  GL_COLOR_CONVERT_PROP_DISABLE_PASSTHROUGH
+};
+
+#define DISABLE_PASSTHROUGH_DAFAULT FALSE
+
 static gboolean gst_gl_color_convert_element_set_caps (GstBaseTransform * bt,
     GstCaps * in_caps, GstCaps * out_caps);
 static GstCaps *gst_gl_color_convert_element_transform_caps (GstBaseTransform *
@@ -54,6 +62,15 @@ static GstFlowReturn gst_gl_color_convert_element_transform (GstBaseTransform *
 static GstCaps *gst_gl_color_convert_element_fixate_caps (GstBaseTransform *
     bt, GstPadDirection direction, GstCaps * caps, GstCaps * othercaps);
 
+static void gst_gl_color_convert_set_property (GObject *object,
+    guint prop_id,
+    const GValue *value,
+    GParamSpec *pspec);
+static void gst_gl_color_convert_get_property (GObject *object,
+    guint prop_id,
+    GValue *value,
+    GParamSpec *pspec);
+
 static GstStaticPadTemplate gst_gl_color_convert_element_src_pad_template =
 GST_STATIC_PAD_TEMPLATE ("src",
     GST_PAD_SRC,
@@ -91,14 +108,18 @@ gst_gl_color_convert_element_class_init (GstGLColorConvertElementClass * klass)
 {
   GstBaseTransformClass *bt_class = GST_BASE_TRANSFORM_CLASS (klass);
   GstElementClass *element_class = GST_ELEMENT_CLASS (klass);
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->set_property = gst_gl_color_convert_set_property;
+  object_class->get_property = gst_gl_color_convert_get_property;
 
   bt_class->transform_caps = gst_gl_color_convert_element_transform_caps;
   bt_class->set_caps = gst_gl_color_convert_element_set_caps;
   bt_class->get_unit_size = gst_gl_color_convert_element_get_unit_size;
   bt_class->filter_meta = gst_gl_color_convert_element_filter_meta;
   bt_class->decide_allocation = gst_gl_color_convert_element_decide_allocation;
-  bt_class->prepare_output_buffer =
-      gst_gl_color_convert_element_prepare_output_buffer;
+  //bt_class->prepare_output_buffer =
+  //    gst_gl_color_convert_element_prepare_output_buffer;
   bt_class->transform = gst_gl_color_convert_element_transform;
   bt_class->stop = gst_gl_color_convert_element_stop;
   bt_class->fixate_caps = gst_gl_color_convert_element_fixate_caps;
@@ -112,6 +133,13 @@ gst_gl_color_convert_element_class_init (GstGLColorConvertElementClass * klass)
       gst_static_pad_template_get
       (&gst_gl_color_convert_element_sink_pad_template));
 
+  g_object_class_install_property (object_class, GL_COLOR_CONVERT_PROP_DISABLE_PASSTHROUGH,
+      g_param_spec_boolean ("disable_passthrough",
+          "Disable passthrough",
+          "Disable passthrough mode",
+          DISABLE_PASSTHROUGH_DAFAULT,
+          G_PARAM_READWRITE));
+
   gst_element_class_set_metadata (element_class,
       "OpenGL color converter", "Filter/Converter/Video",
       "Converts between color spaces using OpenGL shaders",
@@ -123,6 +151,41 @@ gst_gl_color_convert_element_init (GstGLColorConvertElement * convert)
 {
   gst_base_transform_set_prefer_passthrough (GST_BASE_TRANSFORM (convert),
       TRUE);
+  convert->disable_passthrough = FALSE;
+}
+
+static void
+gst_gl_color_convert_set_property (GObject *object,
+    guint prop_id,
+    const GValue *value,
+    GParamSpec *pspec)
+{
+  GstGLColorConvertElement *convert = GST_GL_COLOR_CONVERT_ELEMENT (object);
+  switch (prop_id) {
+    case GL_COLOR_CONVERT_PROP_DISABLE_PASSTHROUGH:
+      convert->disable_passthrough = g_value_get_boolean (value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+static void
+gst_gl_color_convert_get_property (GObject *object,
+    guint prop_id,
+    GValue *value,
+    GParamSpec *pspec)
+{
+  GstGLColorConvertElement *convert = GST_GL_COLOR_CONVERT_ELEMENT (object);
+  switch (prop_id) {
+    case GL_COLOR_CONVERT_PROP_DISABLE_PASSTHROUGH:
+      g_value_set_boolean (value, convert->disable_passthrough);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
 }
 
 static gboolean
@@ -137,6 +200,13 @@ gst_gl_color_convert_element_set_caps (GstBaseTransform * bt,
   if (convert->convert)
     gst_gl_color_convert_set_caps (convert->convert, in_caps, out_caps);
 
+  if(gst_base_transform_is_passthrough (bt) && convert->disable_passthrough){
+    /* if in passthrough mode and disable_passthrough is set to true, 
+     * set passthrough to FALSE*/
+    GST_DEBUG_OBJECT(convert, "Disable passthrough mode");
+    gst_base_transform_set_passthrough(bt, FALSE);
+  }
+
   return TRUE;
 }
 
@@ -233,6 +303,15 @@ static GstFlowReturn
 gst_gl_color_convert_element_transform (GstBaseTransform * bt,
     GstBuffer * inbuf, GstBuffer * outbuf)
 {
+  GstGLColorConvertElement *convert = GST_GL_COLOR_CONVERT_ELEMENT (bt);
+
+  convert->convert->outbuf = outbuf;
+  if (!gst_gl_color_convert_perform (convert->convert, inbuf)) {
+    GST_ELEMENT_ERROR (bt, RESOURCE, NOT_FOUND,
+        ("%s", "Failed to convert video buffer"), (NULL));
+    return GST_FLOW_ERROR;
+  }
+
   return GST_FLOW_OK;
 }
 
diff --git a/ext/gl/gstglcolorconvertelement.h b/ext/gl/gstglcolorconvertelement.h
index 2a0dd1d..5cdbd3a 100644
--- a/ext/gl/gstglcolorconvertelement.h
+++ b/ext/gl/gstglcolorconvertelement.h
@@ -47,6 +47,7 @@ struct _GstGLColorConvertElement
   GstGLColorConvert *convert;
   GstCaps *in_caps;
   GstCaps *out_caps;
+  gboolean disable_passthrough;
 };
 
 struct _GstGLColorConvertElementClass
diff --git a/gst-libs/gst/gl/gstglcolorconvert.c b/gst-libs/gst/gl/gstglcolorconvert.c
index 5915a39..8fa52ab 100644
--- a/gst-libs/gst/gl/gstglcolorconvert.c
+++ b/gst-libs/gst/gl/gstglcolorconvert.c
@@ -681,8 +681,13 @@ _gst_gl_color_convert_set_caps_unlocked (GstGLColorConvert * convert,
   convert->initted = FALSE;
 
   /* If input and output are identical, pass through directly */
-  convert->passthrough =
-      _gst_gl_color_convert_can_passthrough (&in_info, &out_info);
+
+  /* We may disable passthrough via an external property
+   * By the way, when glconvertelement is in passthrough mode, 
+   * the plugin will not call gst_gl_color_convert_perform().*/
+
+  //convert->passthrough =
+      //_gst_gl_color_convert_can_passthrough (&in_info, &out_info);
 #ifndef GST_DISABLE_GST_DEBUG
   if (G_UNLIKELY (convert->passthrough))
     GST_DEBUG_OBJECT (convert,
@@ -1712,18 +1717,20 @@ _do_convert (GstGLContext * context, GstGLColorConvert * convert)
   gint views, v;
   GstVideoOverlayCompositionMeta *composition_meta;
 
-  convert->outbuf = NULL;
+  //convert->outbuf = NULL;
 
   if (!_init_convert (convert)) {
     convert->priv->result = FALSE;
     return;
   }
 
-  convert->outbuf = gst_buffer_new ();
-  if (!gst_gl_memory_setup_buffer (convert->context, NULL, &convert->out_info,
-          NULL, convert->outbuf)) {
-    convert->priv->result = FALSE;
-    return;
+  if (!convert->outbuf) {
+    convert->outbuf = gst_buffer_new ();
+    if (!gst_gl_memory_setup_buffer (convert->context, NULL, &convert->out_info,
+            NULL, convert->outbuf)) {
+      convert->priv->result = FALSE;
+      return;
+    }
   }
 
   if (GST_VIDEO_INFO_MULTIVIEW_MODE (in_info) ==
-- 
1.7.9.5

